qiskit/synthesis/arithmetic/multipliers/hrs_cumulative_multiplier.py

```diff
@@ -0,0 +1,102 @@
+
+"""Compute the product of two qubit registers using classical multiplication approach."""
+
+from __future__ import annotations
+
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.circuit.quantumregister import QuantumRegister
+
+
+def multiplier_cumulative_h18(
+    num_state_qubits: int, num_result_qubits: int | None = None
+) -> QuantumCircuit:
+    r"""A multiplication circuit to store product of two input registers out-of-place.
+
+    The circuit uses the approach from Ref. [1]. As an example, a multiplier circuit that
+    performs a non-modular multiplication on two 3-qubit sized registers is:
+
+    .. plot::
+        :include-source:
+
+        from qiskit.synthesis.arithmetic import multiplier_cumulative_h18
+
+        num_state_qubits = 3
+        circuit = multiplier_cumulative_h18(num_state_qubits)
+        circuit.draw("mpl")
+
+    Multiplication in this circuit is implemented in a classical approach by performing
+    a series of shifted additions using one of the input registers while the qubits
+    from the other input register act as control qubits for the adders.
+
+    Args:
+        num_state_qubits: The number of qubits in either input register for
+            state :math:`|a\rangle` or :math:`|b\rangle`. The two input
+            registers must have the same number of qubits.
+        num_result_qubits: The number of result qubits to limit the output to.
+            If number of result qubits is :math:`n`, multiplication modulo :math:`2^n` is performed
+            to limit the output to the specified number of qubits. Default
+            value is ``2 * num_state_qubits`` to represent any possible
+            result from the multiplication of the two inputs.
+
+    Raises:
+        ValueError: If ``num_result_qubits`` is given and not valid, meaning not
+            in ``[num_state_qubits, 2 * num_state_qubits]``.
+
+    **References:**
+
+    [1] Häner et al., Optimizing Quantum Circuits for Arithmetic, 2018.
+    `arXiv:1805.12445 <https://arxiv.org/pdf/1805.12445.pdf>`_
+
+    """
+    if num_result_qubits is None:
+        num_result_qubits = 2 * num_state_qubits
+    elif num_result_qubits < num_state_qubits or num_result_qubits > 2 * num_state_qubits:
+        raise ValueError(
+            f"num_result_qubits ({num_result_qubits}) must be in between num_state_qubits "
+            f"({num_state_qubits}) and 2 * num_state_qubits ({2 * num_state_qubits})"
+        )
+
+    # define the registers
+    qr_a = QuantumRegister(num_state_qubits, name="a")
+    qr_b = QuantumRegister(num_state_qubits, name="b")
+    qr_out = QuantumRegister(num_result_qubits, name="out")
+
+    circuit = QuantumCircuit(qr_a, qr_b, qr_out)
+
+    # prepare adder as controlled gate
+    # pylint: disable=cyclic-import
+    from qiskit.circuit.library.arithmetic import AdderGate, ModularAdderGate
+
+    adder = AdderGate(num_state_qubits)
+    controlled_adder = adder.control(annotated=True)
```

Comment:
This is also really cool! Using a generic `AdderGate` allows users to choose different synthesis methods via the high-level-synthesis config, without having to change the multiplier gate definition. Using annotated controls opens a door to more circuit optimizations that we can make.